tests/Unit/DataStructures/VectorImplTestHelper.hpp

```diff
@@ -366,5 +376,338 @@ void vector_test_math_after_move(
+
+// struct used for determining the full number of elements in a vector, array of
+// vectors. Designed for use with `test_element_wise_function`
+struct VectorOrArraySize {
+  // array of vectors version, assumes all vectors in the array are the same
+  // size, as will be the case for the relevant test helpers in this detail
+  // namespace
+  template <typename T, size_t S>
+  size_t operator()(const std::array<T, S>& container) const noexcept {
+    return S * VectorOrArraySize{}(container[0]);
+  }
+  // vector version
+  template <typename T,
+            Requires<cpp17::is_arithmetic_v<typename T::ElementType>> = nullptr>
+  size_t operator()(const T& container) const noexcept {
+    return container.size();
+  }
+};
+
+// struct used for obtaining an indexed value in a vector, array of vectors.
+// Designed for use with `test_element_wise_function`
+struct VectorOrArrayAt {
+  // array of vectors version
+  template <typename T, size_t S>
+  auto operator()(std::array<T, S>& container, size_t index) noexcept {
+    return VectorOrArrayAt{}(gsl::at(container, index % S), index / S);
+  }
+  template <typename T, size_t S>
+  auto operator()(const std::array<T, S>& container, size_t index) noexcept {
+    return VectorOrArrayAt{}(gsl::at(container, index % S), index / S);
+  }
+  // vector version
+  template <typename T,
+            Requires<cpp17::is_arithmetic_v<typename T::ElementType>> = nullptr>
+  auto operator()(T& container, size_t index) noexcept {
+    return container.at(index);
+  }
+};
+
+// given an explicit template parameter pack `Wraps`, wrap the elements of
+// `operand`, passed by pointer, element-by-element, and return the
+// resulting tuple of wrapped elements.
+template <typename... Wraps, typename... Operands, size_t... Is>
+auto wrap_tuple(std::tuple<Operands...>& operand_values,
+                std::index_sequence<Is...> /*meta*/) noexcept {
+  return std::make_tuple(wrap(Wraps{}, get<Is>(operand_values))...);
+}
+
+// given the set of types of operands to test (`Operands`), and a set of
+// reference wrappers (`Wraps`), make each operand with random values according
+// to the bound from `Bounds`. Then, call
+// `CHECK_CUSTOM_ELEMENT_WISE_FUNCTION_APPROX` to test the element wise `func`.
+template <typename Function, typename... Bounds, typename... Wraps,
+          typename... Operands, size_t... Is>
+void test_function_on_vector_operands(
+    const Function& function, const std::tuple<Bounds...>& bounds,
+    const std::tuple<Wraps...>& /*wraps*/,
+    const std::tuple<Operands...>& /*operands*/,
+    std::index_sequence<Is...> /*meta*/) noexcept {
+  MAKE_GENERATOR(generator);
+  UniformCustomDistribution<size_t> size_distribution{2, 5};
+  const DataVector used_for_size{size_distribution(generator)};
+  // using each distribution, generate a value for the appropriate operand type
+  // and put it in a tuple.
+  auto operand_values = std::make_tuple(make_with_random_values<Operands>(
```

Comment:
Can these variables be `const`?

Reply:
Afraid not these ones. Trial and error indicates that these propagate to consts inside the wrappers, screwing things up for modifying operators like `+=`.